### PR TITLE
Type field fixes in central management 

### PR DESCRIPTION
**Pull request by fearful-symmetry**
## What does this PR do?

This is a fix for https://github.com/elastic/elastic-agent/issues/1807 . It fixes two different but related issues:
- The `index` field was generated with the top-level `data_stream` field, and not the stream-level field, which produced the wrong index string.
- There's an issue where not all input configs have a stream-level type field; this adds a fix to filebeat and auditbeat to manuall add the type field to the stream based on the top-level input `type`.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: x-pack/libbeat/management/generate.go [hunk 1]
```diff
@@ -171,8 +175,8 @@ func injectStreamProcessors(expected *proto.UnitExpectedConfig, inputType string
 	// for reasons I can't understand, as it just ends up shuffling it around
 	// to individual metricsets anyway, at least on metricbeat
 	if expectedID := expected.GetId(); expectedID != "" {
-		inputId := generateAddFieldsProcessor(mapstr.M{"input_id": expectedID}, "@metadata")
-		processors = append(processors, inputId)
+		inputID := generateAddFieldsProcessor(mapstr.M{"input_id": expectedID}, "@metadata")
```
Comment (cmacknz) [inline x-pack/libbeat/management/generate.go:178]: The comment on this is possibly obsolete because the AST transformations were removed in V2. Do we need to update the comment to reference what V2 currently does? Remember that this code will never interact with the V1 agent.

```go
	// the AST injects input_id at the input level and not the stream level,
	// for reasons I can't understand, as it just ends up shuffling it around
	// to individual metricsets anyway, at least on metricbeat
```

File: x-pack/libbeat/management/generate_test.go [hunk 2]
```diff
@@ -130,6 +130,20 @@ func TestOutputGen(t *testing.T) {
 
 }
 
+func TestOutputIndex(t *testing.T) {
+	inputType := "unused"
```
Comment (cmacknz) [inline x-pack/libbeat/management/generate_test.go:134]:
```suggestion
	dataStreamType := "unused"
```

**Review by cmacknz — COMMENTED** [review]
I think this will fix things. Have you confirmed that starting Filebeat under agent with an input other than `filestream` or `logs` works as expected?

The `winlog` input reproduced this problem reliably: https://github.com/elastic/elastic-agent/issues/1800

Winlog should only run on Windows though, but I think this problem would occur with any non `log` or `filestream` input type like httpjson that is quite widely used: https://github.com/elastic/elastic-agent/blob/8aae0d629e3765fb2fb42e4e111df17a895a6745/specs/filebeat.spec.yml#L94

File: x-pack/libbeat/management/generate.go [hunk 3]
```diff
@@ -142,7 +146,7 @@ func injectIndexStream(expected *proto.UnitExpectedConfig, inputType string, str
 	}
 
 	namespace := DefaultNamespaceName
-	if testNamespace := expected.GetDataStream().GetNamespace(); testNamespace != "" {
+	if testNamespace := streamExpected.GetDataStream().GetNamespace(); testNamespace != "" {
```
Comment (cmacknz) [inline x-pack/libbeat/management/generate.go:149]: I think we may still be missing cases with this parsing. I think the precedence order needs to be:

1. If no data stream information is specified, the default should be `$defaultType-generic-default` which will typically be one of `logs-generic-default` or `metrics-generic-default` depending on the beat. This is described in the data stream naming scheme [blog post](https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme), where the `type` has no default, `dataset` defaults to `generic`, and `namespace` defaults to `default`.
2. Look at the input level `data_stream` field, and override the defaults above with any values specified there. Almost always the `namespace` is specified at this level, and I believe users can edit this for some integrations so we need to consider it.
3. Look at the stream level `data_stream` field, and take any values specified there as the final values.

Here's a few examples of I think this should work:
```yaml
- id: input1             # Sets data stream fields to logs-generic-default
  type: synthetics/icmp
  streams:
    - id: stream1
- id: input1             # Sets data stream fields to synthetics-icmp-default
  type: synthetics/icmp
  data_stream:
    namespace: default
  streams:
    - id: stream1
      data_stream:
        dataset: icmp
        type: synthetics
- id: input1             # Sets data stream fields to synthetics-example-namespace
  type: synthetics/icmp
  data_stream:
    namespace: namespace
    dataset: example
  streams:
    - id: stream1
      data_stream:
        type: synthetics
```

Comment (cmacknz) [reply x-pack/libbeat/management/generate.go:149]: This is what the V1 agent was doing in 8.5, which is almost what I described above except it doesn't explicitly pull the `type` field out of `input.streams.stream[n].data_stream.type` it takes it as an input parameter: https://github.com/elastic/elastic-agent/blob/0a6c4f05e33ed054ce54409a59f6d7da52e07cd2/internal/pkg/agent/transpiler/rules.go#L587-L634

Also the `inject_index` rule wasn't used in every Beat spec file in V1, notably Heartbeat doesn't use it.

Comment (cmacknz) [reply x-pack/libbeat/management/generate.go:149]: Clearly we need a spec on how to parse an agent policy because this is very inconsistent.

Comment (cmacknz) [reply x-pack/libbeat/management/generate.go:149]: Heartbeat does its own parsing here: https://github.com/elastic/beats/blob/fd82372d4b0f520bdafa9338537e48b9d7bb7967/heartbeat/monitors/factory.go#L284-L303

File: x-pack/filebeat/cmd/agent.go [hunk 4]
```diff
@@ -19,10 +19,17 @@ func filebeatCfg(rawIn *proto.UnitExpectedConfig, agentInfo *client.AgentInfo) (
 		return nil, fmt.Errorf("error creating input list from raw expected config: %w", err)
 	}
 
+	for iter := range modules {
+		if _, ok := modules[iter]["type"]; !ok {
+			modules[iter]["type"] = rawIn.Type
```
[the remark this answers is not in the document]
Comment (fearful-symmetry) [reply x-pack/filebeat/cmd/agent.go:24]: Mostly because I'm paranoid. When I was testing, I ran into a lot of issues where the field was sometimes there and sometimes not, so I'd rather be safe.

File: x-pack/heartbeat/cmd/root.go [hunk 5]
```diff
@@ -23,13 +24,22 @@ var RootCmd *cmd.BeatsRootCmd
 // heartbeatCfg is a callback registered via SetTransform that returns a Elastic Agent client.Unit
 // configuration generated from a raw Elastic Agent config
 func heartbeatCfg(rawIn *proto.UnitExpectedConfig, agentInfo *client.AgentInfo) ([]*reload.ConfigWithMeta, error) {
-	//grab and properly format the input streams
-	inputStreams, err := management.CreateInputsFromStreams(rawIn, "metrics", agentInfo)
+	modules, err := management.CreateInputsFromStreams(rawIn, "metrics", agentInfo)
```
[the remark this answers is not in the document]
Comment (fearful-symmetry) [reply x-pack/heartbeat/cmd/root.go:27]: Hmm, alright, good catch.

File: x-pack/heartbeat/cmd/root.go [hunk 6]
```diff
@@ -23,13 +24,22 @@ var RootCmd *cmd.BeatsRootCmd
...
 	if err != nil {
-		return nil, fmt.Errorf("error generating new stream config: %w", err)
+		return nil, fmt.Errorf("error creating input list from raw expected config: %w", err)
 	}
 
-	configList, err := management.CreateReloadConfigFromInputs(inputStreams)
+	// Extract the stream-level type from the input
+	typeField := strings.Split(rawIn.Type, "/")[1]
```
[the remark this answers is not in the document]
Comment (fearful-symmetry) [reply x-pack/heartbeat/cmd/root.go:33]: Ah, good idea.

File: x-pack/libbeat/management/generate.go (hunk 3 above)
Comment (fearful-symmetry) [reply x-pack/libbeat/management/generate.go:149]: > Clearly we need a spec on how to parse an agent policy because this is very inconsistent.

Agreed. I ran into a lot of issues while testing this because of how allover some of the integrations are, this whole system could be much more formalized.

File: x-pack/heartbeat/cmd/root.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (fearful-symmetry) [reply x-pack/heartbeat/cmd/root.go:27]: Re, `synthetics`, I thought the index fields had to be one of `logs` or `metrics` ?

[the remark this answers is not in the document]
Comment (cmacknz) [reply x-pack/heartbeat/cmd/root.go:27]: Maybe it should be, but the `type` part of the datastream is definitely being set to `synthetics` in some cases: https://github.com/elastic/beats/blob/fd82372d4b0f520bdafa9338537e48b9d7bb7967/heartbeat/monitors/factory.go#L296

We have to keep the existing behaviour here to avoid breaking things.

File: x-pack/libbeat/management/generate.go (hunk 1 above)
Comment (fearful-symmetry) [reply x-pack/libbeat/management/generate.go:178]: Yah, this was a reference to the V1 behavior I was trying to emulate, might want to make that a bit more clear.

**Comment by fearful-symmetry** [comment]
Going through all the different input configs I've seen while debugging this, it's making me a tad paranoid and I want to think about a more defensive way to do this, since the `data_stream` and `type` field are a lot less predictable than I anticipated. 

Agreed that we should have a more formalized spec for some of this too.

**Comment by cmacknz** [comment]
> type field are a lot less predictable than I anticipated.

The type field controls which binary the agent starts, so it is predictible in that it always has to be there. The thing to watch out for is that the agent policy supports aliases for several input types, for example there are three ways to specify that the Filebeat `log` input should be started: https://github.com/elastic/elastic-agent/blob/8aae0d629e3765fb2fb42e4e111df17a895a6745/specs/filebeat.spec.yml#L115

```yml
  - name: log
    aliases:
      - logfile
      - event/file
```

**Comment by fearful-symmetry** [comment]
@cmacknz yah, I'm referring to the stream-level `type` fields. This could be a potentially be a quirk/bug in the heartbeat integrations themselves though, since it looks like some heartbeat integrations have the `type` field in the stream, and some don't.

**Comment by cmacknz** [comment]
> I'm referring to the stream-level type fields. This could be a potentially be a quirk/bug in the heartbeat integrations themselves though, since it looks like some heartbeat integrations have the type field in the stream, and some don't.

My understanding is that we should only care about two `type` fields, the input level `type` which we've already discussed, and the `data_stream.type` field which is optional and can be in either `input.data_stream.type` or `input.streams.stream[n].data_stream.type`. 

If I look at the sample synthetics configuration we were provided, it includes a `type: icmp` at the stream level that is non-standard and we shouldn't have to care about. That is just another piece of integration specific configuration that can freely be placed at the stream level, it's essentially equivalent to the `name` field above it.

```yaml
    streams:
      - id: ccfdfd04-7c1c-4c17-93ef-f0d22ab0a1d5
        name: 8.8.8.8
        type: icmp # <--- I don't think we care about this in the agent
        enabled: true
        data_stream:
          dataset: icmp
          type: synthetics # <--- We do care about this type, since it controls the target index
```

Does that clarify things?

**Comment by fearful-symmetry** [comment]
@cmacknz yah, I'm referring to the `type` field issues found here: https://github.com/elastic/elastic-agent/issues/1807#issuecomment-1329748291 while heartbeat was parsing its own config. This is the type field that I'm worried about stepping on in the individual beat callbacks like this:

```
	// Extract the module name from the stream-level type
	// these types are defined in the elastic-agent's specfiles
	for iter := range modules {
		if _, ok := modules[iter]["type"]; !ok {
			modules[iter]["type"] = rawIn.Type
		}
	}
```

Looking at the integrations, it appears that sometimes the synthetic integrations have the stream-level type field, and sometimes they don't, which is where I assume the errors are coming from in the linked issue. Considering how this kind of inconsistency is apparently possible, I just want to make sure we're not breaking anything.

**Comment by cmacknz** [comment]
Right, I am thinking this from the perspective of parsing the agent policy. There is the second problem of where to put the `type` field in the resulting beat configuration.

Given we have:
```yml
- id: id1
  type: synthetics/icmp # <--- This is mandatory to run under agent
  use_output: default
  data_stream:
    namespace: default
  streams:
    - id: ccfdfd04-7c1c-4c17-93ef-f0d22ab0a1d5
      name: 8.8.8.8
      type: icmp # <--- This is specific to the synthetics integration configuration, likely has the same meaning as the input.type above.
      enabled: true
      data_stream:
        dataset: icmp
        type: synthetic
```

Looking at the reference heartbeat configuration at https://www.elastic.co/guide/en/beats/heartbeat/current/heartbeat-reference-yml.html I believe both the `input.type` and the `input.streams.stream[n].type` field here could both end up in the `heartbeat.monitor.type`. 

The Heartbeat configuration format is completely different than the one for Filebeat, I don't think that just inserting the `type` field unconditionally makes sense. It looks like it would need to go into `heartbeat.monitor.type`. 

I'm not even sure that adding logic to extract type makes sense in the `heartbeatCfg` function, we should just pass through the configuration unmodified. Heartbeat already knows how to parse this I think, it wasn't relying on configuration transformations in V1.

The primary bug for heartbeat is us parsing the `data_stream` fields incorrectly.

File: x-pack/heartbeat/cmd/root.go [hunk 7]
```diff
@@ -23,13 +24,22 @@ var RootCmd *cmd.BeatsRootCmd
...
+
+	for iter := range modules {
+		if _, ok := modules[iter]["type"]; !ok {
```
[the remark this answers is not in the document]
Comment (cmacknz) [reply x-pack/heartbeat/cmd/root.go:36]: Per my comment at the top of the review https://github.com/elastic/beats/pull/33921#issuecomment-1335839001 I'm not sure this makes sense here.

[the remark this answers is not in the document]
Comment (cmacknz) [reply x-pack/heartbeat/cmd/root.go:36]: Yeah Heartbeat is parsing the monitor type internally: https://github.com/elastic/beats/blob/64f98cacc874e225ffb66c0f29741c66ff841636/heartbeat/monitors/stdfields/stdfields.go#L33-L37

We should just pass through the configuration from the integration unmodified.

[the remark this answers is not in the document]
Comment (fearful-symmetry) [reply x-pack/heartbeat/cmd/root.go:36]: Yah, agreed.

**Comment by fearful-symmetry** [comment]
Alright, a few last-minute changes before the weekend; I feel like we're going to be tinkering with the logic for extracting from `data_streams` for a while, though.

File: x-pack/heartbeat/cmd/root.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (cmacknz) [reply x-pack/heartbeat/cmd/root.go]: The packetbeat and auditbeat default type still needs to be updated to from "metrics" to "logs" from what I can see.

File: x-pack/libbeat/management/generate.go [hunk 8]
```diff
@@ -225,31 +226,42 @@ func generateAddFieldsProcessor(fields mapstr.M, target string) mapstr.M {
...
-			unitRaw.DataStream.Namespace = DefaultNamespaceName
-		}
-		if unitRaw.GetDataStream().GetDataset() == "" {
-			unitRaw.DataStream.Dataset = DefaultDatasetName
-		}
+// metadataFromDatastreamValues takes the various data_stream values from across the expected config and returns a set of "good" that can be used to add fields
+// to the final beat config. The underlying logic follows a series of steps:
+// 1) start with a set of default values, including the specified defaultDataStream
+// 2) set the values based on the data_stream fields at the input level (i.e, the UnitExpectedConfig)
+// 3) set the values based on the data_stream fields at the stream level (i.e., the proto.Stream)
```
[the remark this answers is not in the document]
Comment (fearful-symmetry) [reply x-pack/libbeat/management/generate.go:233]: I tried to be a little clever and cram all the cases in by modifying the tests, but I should probably just make separate tests so it's more clear in the code.

**Comment by cmacknz** [comment]
Can we always log the configurations that are created at info level? Or at least the input type (when applicable) and data streams that are extracted from the Unit configurations?

Without this problems in this code are undebuggable from agent diagnostics bundles.

**Comment by fearful-symmetry** [comment]
@cmacknz there's a PR here to aid with the debugability here: https://github.com/elastic/beats/pull/33940

I have another upcoming PR that adds some more debug-level statements about unit config at the level of the elastic-agent-client, which might also help. 

Do you thing it's worth it to add more debug statements about `data_stream` state?

File: x-pack/heartbeat/cmd/root.go [hunk 9]
```diff
@@ -24,7 +24,7 @@ var RootCmd *cmd.BeatsRootCmd
 // configuration generated from a raw Elastic Agent config
 func heartbeatCfg(rawIn *proto.UnitExpectedConfig, agentInfo *client.AgentInfo) ([]*reload.ConfigWithMeta, error) {
 	//grab and properly format the input streams
-	inputStreams, err := management.CreateInputsFromStreams(rawIn, "metrics", agentInfo)
+	inputStreams, err := management.CreateInputsFromStreams(rawIn, "synthetics", agentInfo)
```
Comment (cmacknz) [inline x-pack/heartbeat/cmd/root.go:27]: The only outstanding thing for this PR is to make a similar change for packetbeat and auditbeat, which both default to "logs" instead of "metrics"

https://github.com/elastic/beats/pull/33921#discussion_r1038283527

**Review by cmacknz — APPROVED** [review]
Looks good, thanks!

**Comment by fearful-symmetry** [comment]
So, now realizing that this only works with some inputs but not others, as the syslog integrations has a type `logfile` which doesn't seem to be integrated in filebeat:
```
Error creating runner from config: Error creating input. No such input type exist: 'logfile'
```

Not sure if `logfile` is just weird, or if I'm misunderstanding something in filebeat.

**Comment by cmacknz** [comment]
Hmm, so the agent defines the `logfile` input as an alias of the `log` input in the spec file: https://github.com/elastic/elastic-agent/blob/eaecd0d44014c592128d365f01f40f5f4ed26a29/specs/filebeat.spec.yml#L115

```yaml
  - name: log
    aliases:
      - logfile
      - event/file
```

We have two options:
1) My preferred option which is to have the agent translate the aliases to the expected `log` name the Beat understands.
2) Handle this translation on the Beat side, but since we can add new aliases at any time via the spec it seems like the agent side should be dealing with.

@blakerouse do you have any opinion here?

**Comment by fearful-symmetry** [comment]
Yah, using type `"log"` works.

File: x-pack/heartbeat/cmd/root.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (andrewvc) [reply x-pack/heartbeat/cmd/root.go]: This seems right to me. We may add a "scripted" type in the future, but for now it's just `browser` `http` `tcp` `icmp`

[the remark this answers is not in the document]
Comment (andrewvc) [reply x-pack/heartbeat/cmd/root.go]: I should mention however, that browser monitors output 3 different stream types `browser` `browser_screenshot` and `browser_network`. Does that impact your work here. So, that one input creates three separate outputs.

**Comment by andrewvc** [comment]
I'm a bit confused about the discussion here, it'd help to make sure we are on the same page wrt `data_stream.dataset` vs. `data_stream.type`

For synthetics it should always be:

```
data_stream.type: synthetics
data_stream.dataset: http|icmp|tcp|browser|browser_network|browser_screenshot
```

The `type` in the input yml for synthetics dictates which code executes the config, which for `http|icmp|tcp` matches the data stream type.

For browsers the input yml type of `browser` will result in three separate datasets being written: `browser` `browser_screenshot` 
 and `browser_network`.

**Comment by cmacknz** [comment]
> I'm a bit confused about the discussion here, it'd help to make sure we are on the same page wrt data_stream.dataset vs. data_stream.type

@andrewvc thanks for following up. The only thing we explicitly set here is the default `type`, which was mistakenly `metrics` but should have been `synthetics`. This only applies when it isn't specified in the agent policy now.

At a fundamental level most of this PR is addressing bugs in how we parsed the agent input type and data stream fields out of the agent policy. For example see the resolved comment https://github.com/elastic/beats/pull/33921#discussion_r1038323810

We don't actually look at or modify the data stream fields, we are mostly concerned with extracting them properly to configure processors and set the target index name. What is in the integration configuration should be what gets configured at the Beat level.

We can try to set up a quick zoom chat to clarify further if needed, there is already a lot of discussion in this PR making it hard to follow.

**Comment by fearful-symmetry** [comment]
Alright; I fixed the test to use `log` instead of `logfile`. After that, we should be good.

**Comment by cmacknz** [comment]
All tests have passed, we are just waiting for the packaging jobs to complete. Going to merge this to unblock the next 8.6 BC.
